### ISSUE TITLE
Implement parallel reads.

### ISSUE THREAD

**Issue by danielballan** [bluesky/ophyd#754]
From another in-person conversation with @klauer and @tacaswell:

## Opportunity for performance improvement

When a Device is read, it reads its children in serial, issuing a `ReadNotifyRequest' and waiting for the corresponding `ReadNotifyResponse` before proceeding. There is no reason that the requests could not be sent out in parallel, making reads on large Devices more efficient. This will have a real impact: we have seen reports of slow (~1 second) read times that could be addressed by this change.

## Proposed implementation

We could implement this by adding private methods `_read_futures()` and `_read_configuration_future()` to walk the tree and do the reads async-style. The `read()` and `read_configuration()` methods would be refactored to wrap this and return when the future completes.

Note that this approach does not alter any existing API, nor does it any any new (public) API.

## Future possibilities

* We may wish to make this async API to reading public so it can be used directly, rather than wrapped in a synchronous function. We can think on that for a bit, retaining for awhile the freedom to rework/remove/rename `_read_futures()` because it is not _yet_ pubic.
* To improve the performance even more, we could issue the `ReadNotifyRequest` messages in [batches](https://caproto.github.io/caproto/threading-client.html#batched-requests). This seems like could be done inside `_read_futures()` but it would require a lot more work, so it's best to postpone to a second step. It may not provide a very significant speed up compared to the first step, so taking this second step may not  be a very high priority.
